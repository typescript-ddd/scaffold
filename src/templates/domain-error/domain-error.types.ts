import {
  TemplatePropertiesOption,
  TemplatePropertyValue,
  TemplateStringOption,
  TemplateTokenStringOption,
} from "../shared/template";

export type DomainErrorTemplateValues = {
  name: string;
  message: string;
  parameters: TemplatePropertyValue[];
};

export type DomainErrorTemplateOptions = {
  name: TemplateStringOption;
  message: TemplateTokenStringOption;
  parameters: TemplatePropertiesOption;
};
