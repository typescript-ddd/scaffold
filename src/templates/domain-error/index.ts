export * from "./domain-error.generator";
export * from "./domain-error.template";
export * from "./domain-error.types";