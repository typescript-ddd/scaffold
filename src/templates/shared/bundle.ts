import { GenerateContext, Chunk } from "./generate-context";
import { Template, TemplateOption } from "./template";

export interface Bundle<
  TValues extends { [K in keyof TOptions]: any },
  TOptions extends { [key: string]: TemplateOption }
> extends Omit<Template<TValues, TOptions>, "generate"> {
  generate(value: TValues, context: GenerateContext): Chunk[];
}
