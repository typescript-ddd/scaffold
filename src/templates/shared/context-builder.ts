import { GenerateContext, ContextImportDeclaration } from "./generate-context";
import { arrays } from "../../utils";
import { SourceFile } from "ts-morph";

export class ContextBuilder implements GenerateContext {
  private _fileName: string;
  private _importDeclarations: ContextImportDeclaration[] = [];

  private constructor(
    private rootDir: string = "@/src",
    fileName: string = "example.ts"
  ) {
    this._fileName = fileName;
  }

  resolveDir(
    type: "domain" | "infrastructure" | "application" | "presentation",
    subPath?: string
  ): string {
    switch (type) {
      case "domain":
        return `${this.rootDir}/domain${subPath ? `/${subPath}` : ""}`;
      case "infrastructure":
        return `${this.rootDir}/infrastructure${subPath ? `/${subPath}` : ""}`;
      case "application":
        return `${this.rootDir}/application${subPath ? `/${subPath}` : ""}`;
      case "presentation":
        return `${this.rootDir}/presentation${subPath ? `/${subPath}` : ""}`;
      default:
        throw new Error(`Unknown type ${type} for resolveDir`);
    }
  }

  getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[] {
    const modules = this.getModuleSpecifiers();

    const declarations: ContextImportDeclaration[] = [];

    for (const currentModule of modules) {
      const namedImports = this.getNamedImports(currentModule);
      declarations.push({ moduleSpecifier: currentModule, namedImports });
    }

    if (moduleSpecifier) {
      return declarations.filter((x) => x.moduleSpecifier === moduleSpecifier);
    }

    return declarations;
  }

  addImportDeclaration(declaration: ContextImportDeclaration): void {
    this._importDeclarations.push(declaration);
  }

  getModuleSpecifiers(): string[] {
    return arrays.distinct(
      this._importDeclarations.map((x) => x.moduleSpecifier)
    );
  }

  getNamedImports(moduleSpecifier: string): string[] {
    return arrays.distinct(
      this._importDeclarations
        .filter((x) => x.moduleSpecifier === moduleSpecifier)
        .map((x) => x.namedImports)
        .flat()
    );
  }

  writeSourceFileImports(sourceFile: SourceFile) {
    const importDeclarations = this.getImportDeclarations();
    const moduleSpecifiers = this.getModuleSpecifiers();

    const coreImports = importDeclarations.filter(
      (x) => x.moduleSpecifier === "@typescript-ddd/core"
    );

    if (coreImports) {
      sourceFile.addImportDeclaration({
        moduleSpecifier: `@typescript-ddd/core`,
        namedImports: coreImports.map((x) => x.namedImports).flat(),
      });
    }

    // Non core and root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter(
        (x) => x !== "@typescript-ddd/core" && !x.startsWith(this.rootDir)
      )
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }

    // Root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter((x) => x.startsWith(this.rootDir))
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }
  }

  clearImportDeclarations(): void {
    this._importDeclarations = [];
  }

  tokenize(text: string, tokens: { name: string }[]): string {
    let output = text;
    tokens.forEach((token) => {
      output = output.replace(`{{${token.name}}}`, "${" + token.name + "}");
    });
    return "`" + output + "`";
  }

  get fileName(): string {
    return this._fileName;
  }

  static build(rootDir: string): ContextBuilder {
    return new ContextBuilder(rootDir);
  }
}
