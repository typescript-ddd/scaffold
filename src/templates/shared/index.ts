export * from "./context-builder";
export * from "./generate-context";