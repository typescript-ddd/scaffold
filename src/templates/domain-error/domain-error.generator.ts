import { Project } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";
import { DomainErrorTemplateValues } from "./domain-error.types";

export const generateDomainError = (
  values: DomainErrorTemplateValues,
  context: GenerateContext
) => {
  const { name, parameters = [], message = "An error has occurred." } = values;
  const entityName = strings.capitalize(name);
  const className = `${entityName}Error`;

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${className} extends DomainError {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["DomainError"],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(className);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(name)} error.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    parameters: parameters.map((property) => ({
        name: property.name,
        type: property.valueType,
        hasQuestionToken: property.nullable,
        })),
  });
  constructorDeclaration.setBodyText(`super(${context.tokenize(message, parameters)});`);
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      name
    )}.`,
  });

  let output = sourceFile.getFullText();

  return output;
};
