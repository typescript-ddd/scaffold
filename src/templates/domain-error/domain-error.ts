import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";

export type GenerateDomainErrorOptions = {
  name: string;
  message: string;
  parameters?: { name: string, type: string, nullable?: boolean, defaultValue?: string }[];
};

export const generateDomainError = (
  options: GenerateDomainErrorOptions,
  context: GenerateContext
) => {
  const { name, parameters = [], message = "An error has occurred." } = options;
  const entityName = strings.capitalize(name);
  const className = `${entityName}Error`;

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${className} extends DomainError {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["DomainError"],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(className);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(name)} error.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    parameters: parameters.map((property) => ({
        name: property.name,
        type: property.type,
        hasQuestionToken: property.nullable,
        initializer: property.defaultValue,
        })),
  });
  constructorDeclaration.setBodyText(`super(${context.tokenize(message, parameters)});`);
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      name
    )}.`,
  });

  let output = sourceFile.getFullText();

  return output;
};
