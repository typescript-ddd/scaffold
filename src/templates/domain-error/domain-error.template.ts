import { Chunk, GenerateContext } from "../shared";
import { Template } from "../shared/template";
import { generateDomainError } from "./domain-error.generator";
import {
  DomainErrorTemplateOptions,
  DomainErrorTemplateValues,
} from "./domain-error.types";

export class DomainErrorTemplate
  implements Template<DomainErrorTemplateValues, DomainErrorTemplateOptions>
{
  readonly name = "Command";
  readonly description = "Generates a command";
  readonly options: DomainErrorTemplateOptions = {
    name: {
      name: "name",
      label: "Error Name",
      description: "The name of the error",
      optionType: "string",
    },
    message: {
      name: "message",
      label: "Message",
      description: "The default message of the error",
      optionType: "tokenString",
    },
    parameters: {
      name: "parameters",
      label: "Parameters",
      description: "The parameters of the command",
      optionType: "properties",
    },
  };
  readonly defaultValues = {
    name: "UserAlreadyExists",
    message: "User for {{id}} already exists.",
    parameters: [{ name: "id", valueType: "UserId" }],
  };
  generate(
    values: DomainErrorTemplateValues,
    context: GenerateContext,
    chunkName?: string
  ): Chunk {
    return generateDomainError(values, context, chunkName);
  }
}
