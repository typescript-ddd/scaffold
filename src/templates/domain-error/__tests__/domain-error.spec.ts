import { ContextBuilder } from "../../shared";
import { generateDomainError } from "../domain-error";

describe("DomainError", () => {
  it("should generate domain error", () => {
    // Arrange
    const options = {
      name: "UserLockedOut",
      message: "User is locked out.",
    };
    const context = ContextBuilder.build("@src");

    // Act
    const output = generateDomainError(options, context);

    // Assert
    expect(output).toMatchSnapshot();
  });

  it("should generate domain error with properties", () => {
    // Arrange
    const options = {
      name: "UserAlreadyExists",
      message: "User for {{id}} already exists.",
      parameters: [
        { name: "id", type: "UserId"}
      ]
    };
    const context = ContextBuilder.build("@src");
    context.addImportDeclaration({
        moduleSpecifier: context.resolveDir("domain"),
        namedImports: ["UserId"]
    })

    // Act
    const output = generateDomainError(options, context);

    // Assert
    expect(output).toMatchSnapshot();
  });
});
