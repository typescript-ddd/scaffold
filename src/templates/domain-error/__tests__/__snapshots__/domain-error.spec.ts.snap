// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DomainError should generate domain error 1`] = `
"import { DomainError } from "@typescript-ddd/core";

/** Represents a user locked out error. */
export class UserLockedOutError extends DomainError {
    /** Initializes a new instance of a user locked out. */
    constructor() {
        super(\`User is locked out.\`);
    }
}    
"
`;

exports[`DomainError should generate domain error with properties 1`] = `
"import { DomainError } from "@typescript-ddd/core";
import { UserId } from "@src/domain";

/** Represents a user already exists error. */
export class UserAlreadyExistsError extends DomainError {
    /** Initializes a new instance of a user already exists. */
    constructor(id: UserId) {
        super(\`User for \${id} already exists.\`);
    }
}    
"
`;
